### PR TITLE
Reorgnew [WIP]

### PR DESCRIPTION
**Pull request by tomasbrod**

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -3349,282 +3363,341 @@ bool CBlock::ConnectBlock(CTxDB& txdb, CBlockIndex* pindex, bool fJustCheck, boo
...
-        return error("Reorganize() : WriteHashBestChain failed");
+        if (!txdb.TxnCommit())
+            return error("DisconnectBlocksBatch: TxnCommit failed"); /*fatal*/
 
-    // Make sure it's successfully written to disk before changing memory structure
-    if (!txdb.TxnCommit())
-        return error("Reorganize() : TxnCommit failed");
+        // Need to reload all contracts
+        std::string admin_messages;
+        LoadAdminMessages(true, admin_messages);
```
Comment (denravonska) [inline src/main.cpp]: I think the order we want is:
- LoadAdminMessages
- Tally
- GetLifetimeCPID

Tally updates the blocks staked by a CPID and that set dictates which blocks to accumulate interest/research rewards from.

File: src/main.cpp [hunk 2]
```diff
@@ -3349,282 +3363,341 @@ bool CBlock::ConnectBlock(CTxDB& txdb, CBlockIndex* pindex, bool fJustCheck, boo
...
+    if(cnt_dis>0)
     {
-        CBlockIndex* pindex = vConnect[i];
-        CBlock block;
-        if (!block.ReadFromDisk(pindex))
-            return error("Reorganize() : ReadFromDisk for connect failed");
-        if (!block.ConnectBlock(txdb, pindex, false, true))
+
+        //Block was disconnected - User is Re-eligibile for staking
+        StructCPID sMag = GetInitializedStructCPID2(GlobalCPUMiningCPID.cpid,mvMagnitudes);
```
Comment (denravonska) [inline src/main.cpp]: I saw this when I was working with the reorganize and I wonder if we even need it. This resets the last payment of the host, regardless of if it was involved in the reorganized blocks or not.

File: src/main.cpp [hunk 3]
```diff
@@ -3349,282 +3363,341 @@ bool CBlock::ConnectBlock(CTxDB& txdb, CBlockIndex* pindex, bool fJustCheck, boo
...
+        nLastBlockSolved = 0;
+        if (sMag.initialized)
         {
-            // Invalid block
-            return error("Reorganize() : ConnectBlock %s failed", pindex->GetBlockHash().ToString().substr(0,20).c_str());
+            sMag.LastPaymentTime = 0;
```
Comment (denravonska) [inline src/main.cpp:3457]: I saw this when I was working with the reorganize and I wonder if we even need it. This resets the last payment of the host, regardless of if it was involved in the reorganized blocks or not.

File: src/main.cpp [hunk 4]
```diff
@@ -3349,282 +3363,346 @@ bool CBlock::ConnectBlock(CTxDB& txdb, CBlockIndex* pindex, bool fJustCheck, boo
...
-        if (!block.ReadFromDisk(pindex))
-            return error("Reorganize() : ReadFromDisk for disconnect failed");
-        if (!block.DisconnectBlock(txdb, pindex))
-            return error("Reorganize() : DisconnectBlock %s failed", pindex->GetBlockHash().ToString().substr(0,20).c_str());
+        if (!block.ReadFromDisk(pindexBest))
+            return error("DisconnectBlocksBatch: ReadFromDisk for disconnect failed"); /*fatal*/
+        if (!block.DisconnectBlock(txdb, pindexBest))
+            return error("DisconnectBlocksBatch: DisconnectBlock %s failed", pindexBest->GetBlockHash().ToString().c_str()); /*fatal*/
+
+        if (!txdb.WriteHashBestChain(pindexBest->GetBlockHash()))
```
Comment (denravonska) [inline src/main.cpp:3415]: Should we not do this after setting pindexBest to pindexBest->pprev?

Comment (tomasbrod) [reply src/main.cpp:3415]: that is probably an issue :D
WriteHashBestChain must get the block previous to the disconnected one.

**Pull request merged**